Modules/Filtering/FFT/include/itkFFTWComplexToComplexFFTImageFilter.h

```diff
@@ -148,10 +148,80 @@ class ITK_TEMPLATE_EXPORT FFTWComplexToComplexFFTImageFilter : public ComplexToC
 };
 
 
+/** \class FFTWComplexToComplexFFTImageFilterFactory
```

Comment:
Agreed that there is a lot of repetitive code in this change, I'm open to exploring other methods. There were concerns about macro readability in [https://github.com/InsightSoftwareConsortium/ITK/pull/2836](https://github.com/InsightSoftwareConsortium/ITK/pull/2836), would it be viable here though?

A template class could not be used for each factory because we need to register each FFT template class against various pixel types (real or complex, depending on the class) and image dimensions. But this could potentially be accomplished with a macro, yes.

Reply:
I am not sure that it would be worth it to do the necessary template gymnastics, but perhaps worth considering: maybe templates could work instead of macros here.  The part of the generic template that is difficult (e.g. "need to register each FFT template class against various pixel types (real or complex, depending on the class) and image dimensions.") could be implemented by a second template that has the FFT type as its template parameter and which is dedicated to that hard-to-template functionality.  This second template would be an empty class, with specific template instantiations that are non-empty and could be arbitrarily different for each FFT type.

Reply:
That second template would normally be the type of a member within the first template class.  If its whole raison d'être is for one-time registrations, that member might end up being a magic static.

Reply:
@N-Dekker might have another great suggestion of how to reduce code while also avoiding macros 😄

Reply:
A template template would work for passing in FFT classes themselves, but the real issue lies in indicating whether the given FFT template consumes/produces real/complex-valued pixels at compile time. For instance, trying to template `itk::ComplexToComplex1DFFTImageFilter` over real-valued inputs in the templated factory yields a type error:

```
1>------ Build started: Project: ITKFFTTestDriver, Configuration: Release x64 ------
1>itkComplexToComplex1DFFTImageFilterTest.cxx
1>C:\repos\ITK\Modules\Filtering\FFT\include\itkVnlComplexToComplex1DFFTImageFilter.hxx(88,1): error C2664: 'void vnl_fft_1d<double>::bwd_transform(std::vector<std::complex<double>,std::allocator<std::complex<double>>> &)': cannot convert argument 1 from 'VNLVectorType' to 'std::vector<std::complex<double>,std::allocator<std::complex<double>>> &'
1>C:\repos\ITK\Modules\ThirdParty\VNL\src\vxl\core\vnl/algo/vnl_fft_1d.h(61,8): message : see declaration of 'vnl_fft_1d<double>::bwd_transform'

```

This can be resolved with passing in the input and output pixel types explicitly along with the FFT class, but then we have to create and register a separate factory for each FFT filter + pixel type combination which is an extra burden, i.e.:
```
itk::ObjectFactoryBase::RegisterInternalFactoryOnce<itk::FFTImageFilterFactory<itk::VnlComplexToComplex1DFFTImageFilter, std::complex<double>, std::complex<double>>>();
itk::ObjectFactoryBase::RegisterInternalFactoryOnce<itk::FFTImageFilterFactory<itk::VnlComplexToComplex1DFFTImageFilter, std::complex<float>, std::complex<float>>>();
```

Reply:
I think that Lee was suggesting to further delegate that boilerplate to another (internal?) template class.

Reply:
I am thinking more along the lines of the following.

```c++
// This first template is usable across all FFTImageFilters.
template <typename TFFTImageFilter>
class FFTImageFilterFactory : public itk::ObjectFactoryBase
{
public:
  // Easily shared stuff across different TFFTImageFilter types.
  // Note that we can get InputImageType, etc. from TFFTImageFilter.

private:
  // Easily shared stuff ...

  static constexpr FFTImageFilterFactoryHelper<TFFTImageFilter> m_Helper;
};

// Each specific FFTImageFilter will need its own template specialization of only this part
template <typename TFFTImageFilter>
class FFTImageFilterFactoryHelper
{};

template <>
class FFTImageFilterFactoryHelper<FFTWComplexToComplex1DFFTImageFilter>
{
  FFTImageFilterFactoryHelper()
  {
    // Only register those types and dimensions that are correct to register
    // for a FFTWComplexToComplex1DFFTImageFilter.
  }
};

template <>
class FFTImageFilterFactoryHelper<FFTWHalfHermetianForwardFFTImageFilter>
{
  FFTImageFilterFactoryHelper()
  {
    // Only register those types and dimensions that are correct to register 
    // for a FFTWHalfHermetianForwardFFTImageFilter.
  }
};
```

Reply:
Ahhh I see, thank you for the explanation! I think I agree with this route, would at least reduce the factory boilerplate across different FFT filter types.

Reply:
I haven't followed the discussion closely enough to understand the granularity of the factories ... whether it is one factory per template, or separate factories each time the input type or image dimensions varies, etc.  So, this may or may not be sensible: if you can access everything you need for the registrations by looking at the types defined within each possible `TFFTImageFilter` then you won't need the `FFTImageFilterFactoryHelper` template at all.  Furthermore, if there is one type (or more) missing from each possible `TFFTImageFilter` that would be needed for such an approach, that could be remedied by adding one `using` statement (or more) to each.

Big picture though ... I was recommending this template approach for your consideration in case you hadn't already considered it.  But after considering it, feel free to reject it.

Reply:
I think a helper class is both a good and a necessary addition. Each `TFFTImageFilter` will not in itself give us sufficient information regarding valid input/output pixel types. Since we are trying to specialize over several type combinations, the template for the factory class really needs to be:
```
template <template <typename,typename> typename TFFTImageFilter>
class FFTImageFilterFactory : public itk::ObjectFactoryBase
...
```

As a result we cannot get `TInputImage` and `TOutputImage` from `TFFTImageFilter` because it is not yet fully defined, which leaves us little to go on in regards to real or complex pixel types. A helper class will make this much easier to define and to read than a macro or any other method I can think of here.

As you said above, we can add template specializations for `FFTImageFilterFactoryHelper` in each respective FFT class. Overrides need to happen in the factory class, though, and `RegisterOverride` is `protected` in the factory base class. We can address this by making the helper template class a `friend` of `FFTImageFilterFactory`, injecting the factory in the constructor, and calling `OverrideFFTImageFilterType` with input/output pixel types specific to that helper specialization:

```
template <template <typename, typename> typename TFFTImageFilter>
class FFTImageFilterFactory : public itk::ObjectFactoryBase
{

  FFTImageFilterFactory()
  {
    // Specialized helper has the side effect of calling `OverrideFFTImageFilterType`
    // on this factory in its constructor
    static FFTImageFilterFactoryHelper<TFFTImageFilter> helper(this);
  }

  /** Helper must be a friend so that it can register TFFTImageFilter factory
   overrides for valid types (real/complex inputs/outputs) */
  template <template <typename, typename> typename TFFTImageFilter>
  friend class FFTImageFilterFactoryHelper;
};

...

// Factory helper defines static types for initialization
template <>
class FFTImageFilterFactoryHelper<TFFTImageFIlter>
{
public:
  FFTImageFilterFactoryHelper(FFTImageFilterFactory<TFFTImageFIlter> * factory)
  {
    factory->OverrideFFTImageFilterType<std::complex<float>, std::complex<float>, 1>();
...
  }
};
```

Unfortunately this makes compile-time logic a little more complicated to decipher, but it does exactly what we need. I believe it's reasonable to tightly couple `FFTImageFilterFactoryHelper` and `FFTImageFilterFactory` in this way, but I'm certainly open to other opinions.

Reply:
I think that you are saying that using
```c++
template <typename TFFTImageFilter>
class FFTImageFilterFactory : public itk::ObjectFactoryBase
{
  // Easily shareable stuff ...

  // Registration.  Instead of a FFTImageFilterFactoryHelper template, somehow make immediate use of:
  typename TFFTImageFilter::InputImageType::PixelType
  typename TFFTImageFilter::OutputImageType::PixelType
  TFFTImageFilter::InputImageType::ImageDimension
};

```
doesn't work.

Reply:
Correct. Passing `TFFTImageFilter` as a complete type, such as `itk::FFTImageFilterFactory<itk::VnlForwardFFTImageFilter<itk::Image<float,2>,itk::Image<itk::complex<float>,2>>>` requires us to specify pixel types (i.e. `float` vs `double`) and image dimension outside of the factory and likewise creates a separate factory for every combination of types, which is a step back from how factories are currently implemented to handle all of that internally.

Desired behavior is that we can pass in a single, incomplete filter type such as `itk::FFTImageFilterFactory<itk::VnlForwardFFTImageFilter>` and have it handle each of those type combinations internal to a single factory object, which is what we are using the helper to accomplish.

Reply:
I've pushed an update to demonstrate -- take a look at [itkFFTImageFilterFactory.h](https://github.com/InsightSoftwareConsortium/ITK/pull/2864/files#diff-7b78a60a72cb4a113116b7bcc9f48f863618e4984d5eb19ddfc950d69a0d3af8) and [itkComplexToComplex1DFFTImageFilter.h](https://github.com/InsightSoftwareConsortium/ITK/pull/2864/files#diff-d5d04bad66f3a4363e60e0284a98ae70711339036c433b2bd51634fdd0efc4a0) for an example of how factory registration with the helper class can work. 

An example of registering one templated factory for multiple constructor overrides is at [https://github.com/InsightSoftwareConsortium/ITK/pull/2864/files#diff-d9aa9abe766d1781d909247ac146f4cbc855b43c640ab8add376b6bc090b524a](https://github.com/InsightSoftwareConsortium/ITK/pull/2864/files#diff-d9aa9abe766d1781d909247ac146f4cbc855b43c640ab8add376b6bc090b524a).

Reply:
Ah, I see.  You might yet avoid the helper template, with this SFINAE approach:
```c++
template <typename TFFTImageFilter>
class FFTImageFilterFactory : public itk::ObjectFactoryBase
{
  // Usual stuff

  FFTImageFilterFactory()
  {
    // Usual stuff

    // Register this factory with the Register method.  Or maybe call it OverrideFFTImageFilterAllTypes.
    this->Register<typename TFFTImageFilter::InputImageType::PixelType,
                   typename TFFTImageFilter::OutputImageType::PixelType>();
  }

  // Factory registration

  // Only one of these Register methods will be defined, because
  // SFINAE.  First we need the helper is_complex_t.

  template <typename T>
  struct is_complex_t : public std::false_type
  {};

  template <typename T>
  struct is_complex_t<std::complex<T>> : public std::true_type
  {};

  template <typename InputPixelType,
            typename OutputPixelType,
            std::enable_if_t<is_complex_t<InputPixelType>::value, bool> = true,
            std::enable_if_t<is_complex_t<OutputPixelType>::value, bool> = true>
  void
  Register() const
  {
    for (unsigned imageDimension = 1; imageDimension <= 3; ++imageDimension)
    {
      this->OverrideFFTImageFilterType<std::complex<float>, std::complex<float>, imageDimension>();
      this->OverrideFFTImageFilterType<std::complex<double>, std::complex<double>, imageDimension>();
    }
  }

  template <typename InputPixelType,
            typename OutputPixelType,
            std::enable_if_t<!is_complex_t<InputPixelType>::value, bool> = true,
            std::enable_if_t<is_complex_t<OutputPixelType>::value, bool> = true>
  void
  Register() const
  {
    for (unsigned imageDimension = 1; imageDimension <= 3; ++imageDimension)
    {
      this->OverrideFFTImageFilterType<std::float, std::complex<float>, imageDimension>();
      this->OverrideFFTImageFilterType<std::double, std::complex<double>, imageDimension>();
    }
  }

  template <typename InputPixelType,
            typename OutputPixelType,
            std::enable_if_t<is_complex_t<InputPixelType>::value, bool> = true,
            std::enable_if_t<!is_complex_t<OutputPixelType>::value, bool> = true>
  void
  Register() const
  {
    for (unsigned imageDimension = 1; imageDimension <= 3; ++imageDimension)
    {
      this->OverrideFFTImageFilterType<std::complex<float>, float, imageDimension>();
      this->OverrideFFTImageFilterType<std::complex<double>, double, imageDimension>();
    }
  }

  template <typename InputPixelType,
            typename OutputPixelType,
            std::enable_if_t<!is_complex_t<InputPixelType>::value, bool> = true,
            std::enable_if_t<!is_complex_t<OutputPixelType>::value, bool> = true>
  void
  Register() const
  {
    for (unsigned imageDimension = 1; imageDimension <= 3; ++imageDimension)
    {
      this->OverrideFFTImageFilterType<std::float, float, imageDimension>();
      this->OverrideFFTImageFilterType<std::double, double, imageDimension>();
    }
  }
};
```

Reply:
We could probably use [itk::NumericTraits< T >::IsComplex](https://itk.org/Doxygen/html/classitk_1_1NumericTraits.html#ae15d09cd274f758f9505e35e891649ba) instead of `is_complex_t` helper.

Reply:
Interesting, I haven't dealt with SFINAE overrides before, thanks for the suggestion @Leengit . However, if I understand correctly, I think this presents the same problem as before specifically regarding whether or not `TFFTImageFilter` is itself a template.

If the image filter is already specialized:
```cpp
template <typename TFFTImageFilter>
```
Then `TFFTImageFilter::InputImageType` and `TFFTImageFilter::OutputImageType` are valid types, but we cannot override `TFFTImageFilter` for other types as in each `Register` block:
```cpp
    for (unsigned imageDimension = 1; imageDimension <= 3; ++imageDimension)
    {
      // Invalid because TFFTImageFilter is already specialized
      this->OverrideFFTImageFilterType<std::float, float, imageDimension>();
      this->OverrideFFTImageFilterType<std::double, double, imageDimension>();
    }
```

Meanwhile if the image filter is not yet specialized:
```cpp
template <template <typename, typename> typename TFFTImageFilter>
```
Then it is valid to call `OverrideFFTImageFilterType` with various types to register override specializations, but we cannot access `InputImageType` and `OutputImageType` as these are not yet defined:
```cpp
    // Invalid because TFFTImageFilter is not yet specialized
    this->Register<typename TFFTImageFilter::InputImageType::PixelType,
                   typename TFFTImageFilter::OutputImageType::PixelType>();
```

TL;DR: Trying to do everything in one factory class without helpers leads to a chicken-and-the-egg problem, where we are trying to rely on the pixel type to define itself.

Does this line of thinking make sense? Please let me know if I'm misunderstanding.

Reply:
Yes, I think that you are right.  I wasn't fully understanding the implications of the use of 
```c++
template <template <typename, typename> typename TFFTImageFilter>
class FFTImageFilterFactory : public itk::ObjectFactoryBase
```
instead of
```c++
template <typename TFFTImageFilter>
class FFTImageFilterFactory : public itk::ObjectFactoryBase
```

Reply:
Understood, thanks for all your suggestions here @Leengit ! 

I'll move forward with replacing factories with helper specializations for each FFT class.

Reply:
Okay, one more brainstorm.  Probably best to stop reading here, but just in case it floats your boat, we could instead do this in terms of traits, as in 
```c++
template <template <typename, typename> typename TFFTImageFilter>
struct FFTImageFilterTraits
{};

template <>
struct FFTImageFilterTraits<ComplexToComplex1DFFTImageFilter>
{
  using InputPixelFloatType = std::complex<float>;
  using InputPixelDoubleType = std::complex<double>;
  using OutputPixelFloatType = std::complex<float>;
  using OutputPixelDoubleType = std::complex<double>;
};

template <>
struct FFTImageFilterTraits<HalfHermetianForwardFFTImageFilter>
{
  using InputPixelFloatType = std::complex<float>;
  using InputPixelDoubleType = std::complex<double>;
  using OutputPixelFloatType = float;
  using OutputPixelDoubleType = double;
};
```

Reply:
I'm learning a good deal about C++ template possibilities in this thread, thank you 😄 

I like this idea the best, it avoids the added complexity of making the helper a `friend` class and lets underlying pixel type / dimension specifications for factory registration remain within the factory itself. It looks like it also has precedent within ITK already (see `GPUTraits` at [https://github.com/InsightSoftwareConsortium/ITK/blob/master/Modules/Core/GPUCommon/include/itkGPUImage.h#L290](https://github.com/InsightSoftwareConsortium/ITK/blob/master/Modules/Core/GPUCommon/include/itkGPUImage.h#L290)) and isn't particularly difficult to read.

@Leengit I'm going to move forward with this one but feel free to add any more brainstorms that you might have in the meantime 🌩️

Reply:
Here's another question, is it possible for us to access the superclass of an FFT image filter before its template argument list is supplied?

For instance, say we want to register a factory to override `ComplexToComplex1DFFTImageFilter` with `VnlComplexToComplex1DFFTImageFilter`'s constructor. It's reasonable that the `Vnl` class will use the same input/output pixel types as its base class, so we want to define traits for the base class and inherit:
```cpp
template <>
struct FFTImageFilterTraits<ComplexToComplex1DFFTImageFilter>
```
However, `FFTImageFilterTraits<VnlComplexToComplexImageFilter>` fails to match against this signature. Likewise, `VnlComplexToComplexImageFilter::Superclass` fails to compile because the template class is not yet specialized.

A viable workaround is for each derived FFT class to implement its own traits, which is essentially what we would have needed to do with helper classes anyway. But, it would be more convenient and introduce less overhead if traits can be defined at the base level and matched with inherited classes.

Reply:
I'm not sure.  I think at this point you might do it without the inheritance and then we can go back and change that with a follow up pull request if we have any insights.

Reply:
Does this work?  Whatever code was going to use `TFFTImageFilter` as one template parameter to get access to `FFTImageFilterTraits<TFFTImageFilter>::InputPixelType<float>` can also take a second template parameter `TSuperclass` and instead get access to `FFTImageFilterTraits<TSuperclass>::InputPixelType<float>`.

Reply:
This could work, but might get a little clunky as we would need to pass both the derived and base class every time we register a new factory. I'd favor sticking with a new traits object for every derived class, at least for now.

Reply:
@Leengit I've added you as a co-author on the FFT factory commit in this PR, please let me know if there are any other changes you think should be made!